OpenRA.Game/Graphics/CursorProvider.cs

```diff
@@ -26,21 +27,16 @@ public CursorProvider(ModData modData)
 			var sequenceYaml = MiniYaml.Merge(modData.Manifest.Cursors.Select(
 				s => MiniYaml.FromStream(fileSystem.Open(s), s)));
 
-			var shadowIndex = new int[] { };
-
 			var nodesDict = new MiniYaml(null, sequenceYaml).ToDictionary();
-			if (nodesDict.ContainsKey("ShadowIndex"))
-			{
-				Array.Resize(ref shadowIndex, shadowIndex.Length + 1);
-				Exts.TryParseIntegerInvariant(nodesDict["ShadowIndex"].Value,
-					out shadowIndex[shadowIndex.Length - 1]);
-			}
 
-			var palettes = new Dictionary<string, ImmutablePalette>();
-			foreach (var p in nodesDict["Palettes"].Nodes)
-				palettes.Add(p.Key, new ImmutablePalette(fileSystem.Open(p.Value.Value), shadowIndex));
+			var pals = modData.DefaultRules.Actors["world"].TraitInfos<IProvidesCursorPaletteInfo>()
+				.Where(p => p.Palette != null)
+				.ToDictionary(p => p.Palette, p => p);
```

Comment:
This will throw an exception if there are duplicate `p.Palette`s... we probably want to do this manually
```csharp
// Overwrite previous definitions if there are duplicates
var pals = new Dictionary<string, IProvidesCursorPaletteInfo>();
foreach (var p in modData.DefaultRules.Actors["world"].TraitInfos<IProvidesCursorPaletteInfo>())
	if (p.Palette != null)
		pals[p.Palette] = p;
```

Its a shame that LINQ doesn't support something like this directly.